activerecord/lib/active_record/relation.rb

```diff
@@ -985,6 +977,32 @@ def exec_main_query(async: false)
         end
       end
 
+      def _update_each(method_name, id, attributes, &block)
+        # Took from ActiveRecord::Persistence
+        if id.respond_to?(:map) && id.respond_to?(:any?)
+          if id.any?(ActiveRecord::Base)
+            raise ArgumentError,
+              "You are passing an array of ActiveRecord::Base instances to `#{method_name}`. " \
+              "Please pass the ids of the objects by calling `pluck(:id)` or `map(&:id)`."
+          end
```

Comment:
Would be curious to hear other people's opinions on this, but I personally find that mixing in these changes related to validating the input to `update` / `update!` with the changes to address keeping the relation scope are making this more difficult to review. Perhaps we should split these changes out into a separate PR if we think they're warranted?

Reply:
I believe the expectation is that calling `update` or `update!` on the scope should behave the same way as if it would have been called on the model itself but filtering down (or even extending, think about the default scopes) the results using the current scope.

Since the model `update` and `update!` support these arguments and the implementation has these helping errors or input validations, therefore I cannot see how this could be split up differently to also satisfy that expectation.